src/customParser.ts

```diff
@@ -0,0 +1,330 @@
+  }
+
+  readUint8Array(numberOfValues: number) {
+  const newArray = [];
+  for (let i = 0; i < numberOfValues; i++) {
+    newArray.push(this.dataView.getUint8(this.location + i));
+  }
+  this.location += numberOfValues;
+  return newArray;
+  }
+
+  readFloat32Array(numberOfValues: number) {
+  const newArray = [];
+  for (let i = 0; i < numberOfValues; i++) {
+    newArray.push(this.dataView.getFloat32(this.location + i * 4, this.flip));
+  }
+  this.location += numberOfValues * 4;
+  return newArray;
+  }
+
+  skip(numberOfBytes: number) {
+    this.location += numberOfBytes;
+  }
+}
+
+function parseCustomFileFormat(asArrayBuffer: ArrayBuffer, flip: boolean) {
+  const fileReader = new BufferReader(asArrayBuffer, true);
+
+  const infoHeader = fileReader.readUint32Array(4);
+  const magicBytes = infoHeader[0];
+  const formatVersion = infoHeader[1];
+  const optimizerVersion = infoHeader[2];
+  const arrayCount = infoHeader[3];
+
+  const arraysList = ['colors', 'x_values', 'y_values', 'z_values', 'normals', 'deltas', 'heights', 'radiuses',
+  'angles', 'matrixes', 'x_translations', 'y_translations', 'z_translations'];
+  const arraysDictionary: any = {};
+  for (let arrayId = 0; arrayId < arrayCount;  arrayId++) {
+    const clusterCount = fileReader.readUint32();
+    const bytesForOneValue = fileReader.readUint32();
+    const type = arraysList[arrayId];
+
+    if ((type === 'colors') && (bytesForOneValue === 4)) {
+      const colorValues = fileReader.readUint8Array(clusterCount * 4);
+      arraysDictionary[type] = [];
+      for (let j = 0; j < clusterCount; j++) {
+        arraysDictionary[type].push([]);
+        for (let k = 0; k < 4; k++) {
+          arraysDictionary[type][j].push(colorValues[j * 4 + k]);
+        }
+      }
+    } else if ((type === 'normals') && (bytesForOneValue === 12)) {
+      const normalValues = fileReader.readFloat32Array(clusterCount * 3);
+      arraysDictionary[type] = [];
+      for (let j = 0; j < clusterCount; j++) {
+        const newNormal = new THREE.Vector3(normalValues[j * 3], normalValues[j * 3 + 1], normalValues[j * 3 + 2]);
+        arraysDictionary[type].push(newNormal);
+      }
+    } else if (bytesForOneValue === 4) {
+      arraysDictionary[type] = fileReader.readFloat32Array(clusterCount);
+    } else {
+      throw Error('Well this is unexpected. bytesForOneValue is ' + bytesForOneValue);
+    }
+  }
+
+  const geometryAttributes = [];
+  while (true) {
+    const geometryType = fileReader.readUint32();
+    const geometryCount = fileReader.readUint32();
+    const attributeCount = fileReader.readUint32();
+    const byteCount = fileReader.readUint32();
+
+    const indexes = new FibonacciDecoder(asArrayBuffer, fileReader.location, byteCount);
+    fileReader.skip(byteCount);
+
+    geometryAttributes.push({'type': geometryType, 'indexes': indexes, 'count': geometryCount,
+    'attributeCount': attributeCount});
+
+    if (fileReader.location === asArrayBuffer.byteLength) {
```

Comment:
What happens if we by mistake have a location larger than the byteLength? Is it possible if there is an error in the file? Instead of infinite loop, we could maybe have >= test?